src/libsyntax/parse/parser.rs

```diff
@@ -3570,7 +3570,37 @@ impl<'a> Parser<'a> {
                     err.emit();
                 }
             } else if !self.check(&token::CloseDelim(token::Paren)) {
-                fields.push(self.parse_pat(None)?);
+                let field_pat = self.parse_pat(None)?;
```

Comment:
This is the real crux of the patch set. I tried implementing `or-pattern` parsing a few different ways, but settled on implementing it in the parsing of a list. So `<pat> | <pat>` is not a valid pattern, but `(<pat> | <pat>)` is. This means that the pattern `Foo { bar: 0 | 1 }` would not be a valid pattern, but `Bar(0 | 1)` is.

Reply:
So not parsing `Foo { bar: 0 | 1 }` as a valid pattern would be incorrect from the RFCs POV.

I would also caution against attempting to parse what naturally fits as a binary operator as a list inside a parser. If you want to deal with lists, I would linearlize the tree in lowering instead.

An or-pattern should be the lowest binding operator in the pattern grammar, so you should be able to replace `parse_pat_with_range_pat` in `fn parse_pat` as the thing you attempt to parse first.

Reply:
Also, I'd add some tests to see that you are parsing or-patterns as per the RFC and using `parse_top_pat` in the right places.

Reply:
If this were a binary operator, we wouldn't accept the leading separator `| A | B | C`.
With the leading separator it's a list.

Reply:
The leading separator is only permitted at the top level (i.e. no `Some( | p | q)`) and is easily dealt with at the top level in the same manner as done in `fn parse_pats` before entering the regular pattern grammar.

Reply:
Seems like the implemented `(A | B)` is a strict subset of the full syntax, so in theory it can be merged if we want to work incrementally and land fast.

Reply:
Sure; up to @dlrobertson how they want to stage the implementation imo. :)

Reply:
@Centril good points. I posted this PR early to get feedback on the parsing, so this is exactly what I'm looking for. Depending on the level of effort required to implement non-list or-patterns I might hold off on implementing it. As @petrochenkov mentioned this is a fairly "easy" subset of the RFC to parse/support that also covers many of the use-cases for or-patterns.

Reply:
Sure. I do want to note that while implementation staging is fine in terms of the work towards supporting or-patterns, I do not want to stabilize an easy subset like this.

Reply:
> I do not want to stabilize an easy subset like this.

Agreed.